nemo_curator/image/embedders/timm.py

```diff
@@ -0,0 +1,132 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+from typing import Iterable
+
+import nvidia.dali.fn as fn
+import nvidia.dali.types as types
+import timm
+import torch
+from nvidia.dali import pipeline_def
+from nvidia.dali.plugin.pytorch import feed_ndarray
+
+from nemo_curator.image.embedders.base import ImageEmbedder
+from nemo_curator.utils.image.transforms import convert_transforms_to_dali
+
+
+class TimmImageEmbedder(ImageEmbedder):
+    def __init__(
+        self,
+        model_name: str,
+        pretrained: bool = False,
+        batch_size: int = 1,
+        num_threads_per_worker: int = 4,
+        image_embedding_column: str = "image_embedding",
+        normalize_embeddings: bool = True,
+        classifiers: Iterable = [],
+        autocast: bool = True,
+    ) -> None:
+        super().__init__(
+            model_name=model_name,
+            image_embedding_column=image_embedding_column,
+            classifiers=classifiers,
+        )
+        self.pretrained = pretrained
+        self.batch_size = batch_size
+        self.num_threads_per_worker = num_threads_per_worker
+        self.normalize_embeddings = normalize_embeddings
+        self.autocast = autocast
+
+        # Load the model to get the transforms
+        model = timm.create_model(self.model_name, pretrained=self.pretrained)
+        torch_transforms = timm.data.create_transform(
+            **timm.data.resolve_data_config(model.pretrained_cfg)
+        )
+        self.dali_transforms = convert_transforms_to_dali(torch_transforms)
+
+    def load_dataset_shard(self, tar_path: str, device_id=0):
+        # Create the DALI pipeline
+        @pipeline_def(
+            batch_size=self.batch_size,
+            num_threads=self.num_threads_per_worker,
+            device_id=device_id,
+        )
+        def webdataset_pipeline(_tar_path: str):
+            img_raw, text, json = fn.readers.webdataset(
+                paths=_tar_path,
+                ext=["jpg", "txt", "json"],
+                missing_component_behavior="error",
+            )
+            img = fn.decoders.image(img_raw, device="mixed", output_type=types.RGB)
+
+            for transform in self.dali_transforms:
+                img = transform(img)
+
+            return img, text, json
+
+        pipeline = webdataset_pipeline(tar_path)
+        pipeline.build()
+
+        total_samples = pipeline.epoch_size()
+        total_samples = total_samples[list(total_samples.keys())[0]]
+
+        samples_completed = 0
+        while samples_completed < total_samples:
+            image, text, meta = pipeline.run()
+            image = image.as_tensor()
+
+            image_torch = torch.empty(
+                image.shape(), dtype=torch.float32, device=f"cuda:{device_id}"
```

Comment:
I would rather have us default to `cuda` .

Reply:
I think we should just get rid of `device_id` based logic to keep the code simpler.

Reply:
So you think I should hardcode `device_id=0` [here](https://github.com/NVIDIA/NeMo-Curator/pull/227/files#diff-98a23453d305cf68c670ff378441ddb00d871c11e1ad0212e02636d87da9afaaR65), remove it as a parameter from the function, and use "cuda" everywhere else?

Reply:
Yeah, exactly that. I think we should default to what Pytorch behavior is . I am afraid of folks with non default Pytorch context/setup might run into weird issues that will be hard to debug otherwise.